packages/wrangler/src/cloudchamber/images/list.ts

```diff
@@ -0,0 +1,248 @@
+	}
+}
+
+async function ListTags(
+	responses: TagsResponse[],
+	digests: boolean = false,
+	json: boolean = false
+) {
+	if (!digests) {
+		responses = responses.map((resp) => {
+			return {
+				name: resp.name,
+				tags: resp.tags.filter((t) => !t.startsWith("sha256")),
+			};
+		});
+	}
+	// Remove any repos with no tags
+	responses = responses.filter((resp) => {
+		return resp.tags !== undefined && resp.tags.length != 0;
+	});
+	if (json) {
+		logger.log(JSON.stringify(responses, null, 2));
+	} else {
+		const rows = responses
+			.map((resp) => {
+				return {
+					REPOSITORY: resp.name,
+					TAG: resp.tags.join(" "),
+				};
+			})
+			.flat();
+		logger.table(rows);
+	}
+}
+
+async function listTags(repo: string, creds: string): Promise<string[]> {
+	const url = new URL(`https://${domain}`);
+	const baseUrl = `${url.protocol}//${url.host}`;
+	const tagsUrl = `${baseUrl}/v2/${repo}/tags/list`;
+
+	const tagsResponse = await fetch(tagsUrl, {
+		method: "GET",
+		headers: {
+			Authorization: `Basic ${creds}`,
+		},
+	});
+	const tagsData = (await tagsResponse.json()) as TagsResponse;
+	return tagsData.tags || [];
+}
+
+async function listRepos(creds: string): Promise<string[]> {
+	const url = new URL(`https://${domain}`);
+
+	const catalogUrl = `${url.protocol}//${url.host}/v2/_catalog`;
+
+	const response = await fetch(catalogUrl, {
+		method: "GET",
+		headers: {
+			Authorization: `Basic ${creds}`,
+		},
+	});
+	if (!response.ok) {
+		console.log(JSON.stringify(response));
+		throw new Error(
+			`Failed to fetch repository catalog: ${response.status} ${response.statusText}`
+		);
+	}
+
+	const data = (await response.json()) as CatalogResponse;
+
+	return data.repositories || [];
+}
+
+async function deleteTag(
+	baseUrl: string,
+	image: string,
+	tag: string,
+	creds: string
+) {
+	const manifestAcceptHeader =
+		"application/vnd.oci.image.manifest.v1+json, application/vnd.docker.distribution.manifest.v2+json";
+	const manifestUrl = `${baseUrl}/v2/${image}/manifests/${tag}`;
+	// grab the digest for this tag
+	const headResponse = await fetch(manifestUrl, {
+		method: "HEAD",
+		headers: {
+			Authorization: `Basic ${creds}`,
+			Accept: manifestAcceptHeader,
+		},
+	});
+	if (!headResponse.ok) {
+		throw new Error(
+			`failed to retrieve tag info for ${tag}: ${headResponse.status} ${headResponse.statusText}`
+		);
+	}
+
+	const digest = headResponse.headers.get("Docker-Content-Digest");
+	if (!digest) {
+		throw new Error(`Digest not found for tag "${tag}".`);
+	}
+
+	const deleteUrl = `${baseUrl}/v2/${image}/manifests/${tag}`;
+	const deleteResponse = await fetch(deleteUrl, {
+		method: "DELETE",
+		headers: {
+			Authorization: `Basic ${creds}`,
+			Accept: manifestAcceptHeader,
+		},
+	});
+
+	if (!deleteResponse.ok) {
+		throw new Error(
+			`Failed to delete tag "${tag}" (digest: ${digest}): ${deleteResponse.status} ${deleteResponse.statusText}`
+		);
+	}
+}
+
+async function getCreds(): Promise<string> {
```

Comment:
Should we memoize this for the duration of the expiration and only fetch another if the JWT has expired?

Reply:
Yeah. Let us handle this another time, since it should be its own PR.